Region/RegionStats.cc

```diff
@@ -127,51 +143,78 @@ void RegionStats::calcHistogram(int channel, int stokes, int nBins, float minVal
 
     // save for next time
     setHistogram(channel, stokes, histogramMsg);
-    m_histogramsValid = true;
 }
 
 // ***** Statistics *****
 
 void RegionStats::setStatsRequirements(const std::vector<int>& statsTypes) {
-    m_regionStats = statsTypes;
+    m_statsReqs = statsTypes;
 }
 
 size_t RegionStats::numStats() {
-   return m_regionStats.size();
+   return m_statsReqs.size();
 }
 
-void RegionStats::fillStatsData(CARTA::RegionStatsData& statsData, const casacore::SubLattice<float>& subLattice) {
+void RegionStats::fillStatsData(CARTA::RegionStatsData& statsData,
+        const casacore::SubLattice<float>& subLattice, int channel, int stokes) {
     // Fill RegionStatsData with statistics types set in requirements.
-    // Sublattice shape may be empty because of no xyregion (outside image or 0 pixels selected),
-    // or lattice mask removed all NaN values
-    if (m_regionStats.empty()) {  // no requirements set, add empty StatisticsValue
+    if (m_statsReqs.empty()) {  // no requirements set, add empty StatisticsValue
         auto statsValue = statsData.add_statistics();
         statsValue->set_stats_type(CARTA::StatsType::None);
-        return;
     } else {
         std::vector<std::vector<double>> results;
-        // stats for entire region, not per channel
-        bool haveStats(getStatsValues(results, m_regionStats, subLattice, false));
-        // update message whether have stats or not
-        for (size_t i=0; i<m_regionStats.size(); ++i) {
-            // add StatisticsValue to message
-            auto statsValue = statsData.add_statistics();
-            auto statType = static_cast<CARTA::StatsType>(m_regionStats[i]);
-            statsValue->set_stats_type(statType);
-            if (!haveStats || results[i].empty()) { // region outside image or NaNs
-                if (statType==CARTA::NumPixels) {
-                    statsValue->set_value(0.0);
+        if (m_statsValid && (m_statsData.count(stokes)) && (m_statsData.at(stokes).count(channel))) {
+            // used stored stats
+            try {
+                std::vector<double> storedStats(m_statsData.at(stokes).at(channel));
+                for (size_t i=0; i<m_statsReqs.size(); ++i) {
+                    // add StatisticsValue to message
+                    auto statsValue = statsData.add_statistics();
+                    auto cartaStatType = static_cast<CARTA::StatsType>(m_statsReqs[i]);
+                    statsValue->set_stats_type(cartaStatType);
+                    statsValue->set_value(storedStats[cartaStatType-13]);
```

Comment:
Why 13? Could we add a named constant for this?

Reply:
The enum I got from @markccchiang from CARTA 0.9 had weird numbering scheme...so I'm to blame for this. We can fix this magic number by fixing the ICD.

Reply:
Ah. I have no objection to the *value* of the magic number; I'd just like it not to be magic. :)

Reply:
[cartaStatType - CARTA::NumPixels] is also odd.  It would be clearest in code if the enum started with NumPixels and ended with None, then the index into the vector would be the stat's enum value itself.  A bit uneasy about this whole thing, maybe yet another map (!) would be better?

Reply:
@pford that sounds like a good idea, we can start with NumPixels. Do we even need a value for `None`? We could just default to using `Sum` for single-pixel profiles. That way, if we extend the list of stats, we don't have to move `None` around, or start adding additional types _after_ `None`.

Reply:
The backend does not need type None, does not read the stats type for a point region.  If spectral requirements are set for a Point, the backend just gets the spectral profile data, sets the type in the return message to None, and sends it.  We could use another type (Sum) for both messages.  StatsType::None is not used to turn off profiles; an empty SpectralConfig is sent instead.